src/dotnet/ToolPackage/LockFileMatcher.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Linq;
+
+namespace Microsoft.DotNet.ToolPackage
+{
+    internal class LockFileMatcher
+    {
+        private static string[] _pathInLockFilePathInArray;
```

Comment:
Why is this whole class static?

Reply:
@peterhuene including why static field. It simply look nicer in linq. `Where(p => LockFileMatcher.MatchesFile(p, entry))`. I want to test it separately, and it is not very related to ToolPackageObtainer.cs, so it is a separate class. I don't want to pass _pathInLockFilePathInArray and _entryPointPathInArray all over the place. So I used field.

Do you think make it a non static class it fine? Where(p => new LockFileMatcher(p, entry).MatchesFile()) another minor downside is, there will be extra allocation.

Reply:
I agree that you shouldn't use static state to keep track of the internal execution state of a method.

What I would do is move the static fields inside the method, and move the private methods inside the internal method as [local functions](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions).

Reply:
I see, I think make it non static class can satisfy all

Reply:
To be clear, I do think this should remain a static class; making these instance fields does not address the reentrancy problem.

Reply:
It seems to me like both the fields could simply be passed as parameters to the private static methods, no?

Reply:
This is my plan, it is good?

``` C#
internal class LockFileMatcher
    {
        private readonly string[] _pathInLockFilePathInArray;
        private readonly string[] _entryPointPathInArray;

        public LockFileMatcher(LockFileItem lockFileItem, string targetRelativeFilePath)
        {
            _pathInLockFilePathInArray = SplitPathByDirectorySeparator(lockFileItem.Path);
            _entryPointPathInArray = SplitPathByDirectorySeparator(targetRelativeFilePath);
        }

        internal bool MatchesFile()
        {
            return EntryPointHasFileAtLeast()
                   && PathInLockFileHasExcatlyToolsTfmRidAsParentDirectories()
                   && RestMatchsOtherThanPathInLockFileHasToolsTfmRidAsParentDirectories();
        }
```

Reply:
Or as Livar suggested, make it static method, but downside is passing pathInLockFilePathInArray entryPointPathInArray down to PathInLockFileHasExcatlyToolsTfmRidAsParentDirectories and RestMatchsOtherThanPathInLockFileHasToolsTfmRidAsParentDirectories.

Reply:
I would just do it as a static method with local functions.  Something like this:

```C#
internal class LockFileMatcher
{
    internal static bool MatchesFile(string pathInLockFile, string entryPoint)
    {
        string[] pathInLockFilePathInArray = SplitPathByDirectorySeparator(pathInLockFile);
        string[] entryPointPathInArray = SplitPathByDirectorySeparator(entryPoint);

        return EntryPointHasFileAtLeast()
                && PathInLockFileHasExcatlyToolsTfmRidAsParentDirectories()
                && RestMatchsOtherThanPathInLockFileHasToolsTfmRidAsParentDirectories();

        //  Local functions
        bool RestMatchsOtherThanPathInLockFileHasToolsTfmRidAsParentDirectories()
        {
            string[] pathAfterToolsTfmRid = pathInLockFilePathInArray.Skip(3).ToArray();
            return !pathAfterToolsTfmRid
                .Where((directoryOnEveryLevel, i) => directoryOnEveryLevel != entryPointPathInArray[i])
                .Any();
        }

        bool PathInLockFileHasExcatlyToolsTfmRidAsParentDirectories()
        {
            if (pathInLockFilePathInArray.Length - entryPointPathInArray.Length != 3)
            {
                return false;
            }

            if (pathInLockFilePathInArray[0] != "tools")
            {
                return false;
            }

            return true;
        }

        bool EntryPointHasFileAtLeast()
        {
            return entryPointPathInArray.Length >= 1;
        }

        string[] SplitPathByDirectorySeparator(string path)
        {
            return path.Split('\\', '/');
        }
    }    
}
```